google/cloud/sql/connector/instance_connection_manager.py

```diff
@@ -371,12 +371,40 @@ async def _perform_refresh(self) -> asyncio.Task:
 
         logger.debug("Entered _perform_refresh")
 
-        self._current = self._loop.create_task(self._get_instance_data())
-        # Ephemeral certificate expires in 1 hour, so we schedule a refresh to happen in 55 minutes.
-
-        self._next = self._loop.create_task(self._schedule_refresh())
-
-        return self._current
+        refresh_task = self._loop.create_task(self._get_instance_data())
+
+        async def _refresh_callback(task: asyncio.Task) -> None:
+            try:
+                task.result()
+            except Exception as e:
+                logger.warn(
+                    "An error occurred while performing refresh. Retrying immediately.",
+                    e,
+                )
+                instance_data = None
+                try:
+                    if isinstance(self._current, asyncio.Task):
+                        instance_data = self._current.result()
+                except Exception:
+                    # Current result is invalid, no-op
+                    logger.debug("Current instance data is invalid.")
+                if (
+                    instance_data is None
+                    or instance_data.expiration < datetime.datetime.now()
+                ):
+                    self._current = refresh_task
+                    # TODO: Implement force refresh method and a rate-limiter for perform_refresh
+                    # Retry by scheduling a refresh 60s from now.
```

Comment:
Maybe we should make this PR first

Reply:
I actually think we should do this one first because the connector currently has no retry mechanism. If we get a failed refresh attempt, we're just stuck with it for the next hour :/